pkg/csv2lp/csvTable.go

```diff
@@ -0,0 +1,464 @@
+	for i := 0; i < len(columns); i++ {
+		col := columns[i]
+		switch {
+		case col.Label == labelMeasurement || col.LinePart == linePartMeasurement:
+			t.cachedMeasurement = &col
+		case col.Label == labelTime || col.LinePart == linePartTime:
+			if t.cachedTime != nil && t.cachedTime.Label != labelStart && t.cachedTime.Label != labelStop {
+				log.Printf("WARNING: at most one dateTime column is expected, '%s' column is ignored\n", t.cachedTime.Label)
+			}
+			t.cachedTime = &col
+		case len(strings.TrimSpace(col.Label)) == 0 || col.LinePart == linePartIgnored:
+			// ignored columns that are marked to be ignored or without a label
+		case col.Label == labelFieldName:
+			t.cachedFieldName = &col
+		case col.Label == labelFieldValue:
+			t.cachedFieldValue = &col
+		case col.LinePart == linePartTag:
+			col.escapedLabel = escapeTag(col.Label)
+			t.cachedTags = append(t.cachedTags, col)
+		case col.LinePart == linePartField:
+			col.escapedLabel = escapeTag(col.Label)
+			t.cachedFields = append(t.cachedFields, col)
+		default:
+			if defaultIsField {
+				col.escapedLabel = escapeTag(col.Label)
+				t.cachedFields = append(t.cachedFields, col)
+			}
+		}
+	}
+	// line protocol requires sorted tags
+	if t.cachedTags != nil && len(t.cachedTags) > 0 {
+		sort.Slice(t.cachedTags, func(i, j int) bool {
+			return t.cachedTags[i].Label < t.cachedTags[j].Label
+		})
+	}
+	// setup timezone for timestamp column
+	if t.cachedTime != nil && t.cachedTime.TimeZone == nil {
+		t.cachedTime.TimeZone = t.timeZone
+	}
+
+	t.lpColumnsCached = true // line protocol columns are now fresh
+}
+
+// CreateLine produces a protocol line out of the supplied row or returns error
+func (t *CsvTable) CreateLine(row []string) (line string, err error) {
+	buffer := make([]byte, 100)[:0]
+	buffer, err = t.AppendLine(buffer, row)
+	if err != nil {
+		return "", err
+	}
+	return *(*string)(unsafe.Pointer(&buffer)), nil
+}
+
+// AppendLine appends a protocol line to the supplied buffer and returns appended buffer or an error if any
+func (t *CsvTable) AppendLine(buffer []byte, row []string) ([]byte, error) {
+	if t.computeLineProtocolColumns() {
+		// validate column data types
+		if t.cachedFieldValue != nil && !IsTypeSupported(t.cachedFieldValue.DataType) {
+			return buffer, CsvColumnError{
+				t.cachedFieldValue.Label,
+				fmt.Errorf("data type '%s' is not supported", t.cachedFieldValue.DataType),
+			}
+		}
+		for _, c := range t.cachedFields {
+			if !IsTypeSupported(c.DataType) {
+				return buffer, CsvColumnError{
+					c.Label,
+					fmt.Errorf("data type '%s' is not supported", c.DataType),
+				}
+			}
+		}
+	}
+
+	if t.cachedMeasurement == nil {
+		return buffer, errors.New("no measurement column found")
+	}
+	measurement := t.cachedMeasurement.Value(row)
+	if measurement == "" {
+		return buffer, CsvColumnError{
+			t.cachedMeasurement.Label,
+			errors.New("no measurement supplied"),
+		}
+	}
+	buffer = append(buffer, escapeMeasurement(measurement)...)
```

Comment:
These `append` functions can end up being costly. In this case, I would recommend you leave the function to receive a `buffer []byte` but then copy that to a local `bytes.Buffer` and use the `Write` function to add the characters you need.

https://golang.org/pkg/bytes/#Buffer

Reply:
I also use `appendConverted` fn to append field values or various data types, the `appendConverted` internally uses `strConv.append*` functions that accept and return []byte. A change to `bytes.Buffer` would then require (on the other hand) to create temporary objects for every float, int, uint value that I don't need now. I am not sure that the situation gets better after this change. So it is trade-off situation herein. I would rather keep it the way it is.

I wish to have `append` optimized by the compiler or `bytes.Buffer` also accepting primitive types for writing.